sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TableOutputResolver.scala

```diff
@@ -74,6 +66,42 @@ object TableOutputResolver {
     }
   }
 
+  private def reorderColumnsByName(
+      inputCols: Seq[NamedExpression],
+      expectedCols: Seq[Attribute],
+      conf: SQLConf,
+      addError: String => Unit,
+      colPath: Seq[String] = Nil): Seq[NamedExpression] = {
+    expectedCols.flatMap { expectedCol =>
+      val matched = inputCols.filter(col => conf.resolver(col.name, expectedCol.name))
+      val newColPath = colPath :+ expectedCol.name
+      if (matched.isEmpty) {
+        addError(s"Cannot find data for output column '${newColPath.quoted}'")
+        None
+      } else if (matched.length > 1) {
+        addError(s"Ambiguous column name in the input data: '${newColPath.quoted}'")
+        None
+      } else {
+        val matchedCol = matched.head match {
+          case a: Attribute => a.withName(expectedCol.name)
+          case a: Alias => a.withName(expectedCol.name)
+          case other => other
+        }
+        (matchedCol.dataType, expectedCol.dataType) match {
+          case (input: StructType, expected: StructType) =>
+            val fields = input.zipWithIndex.map { case (f, i) =>
+              Alias(GetStructField(matchedCol, i, Some(f.name)), f.name)()
+            }
+            val reordered = reorderColumnsByName(
+              fields, expected.toAttributes, conf, addError, newColPath)
+            Some(Alias(CreateStruct(reordered), expectedCol.name)())
+          case _ =>
```

Comment:
@cloud-fan Thanks for the fix !! I am thinking, do we need cases for Array and Map types ? Like for cases.
Struct<a: Array<Struct>> . Don't we need to create the arrays and maps from the remapped inner structs ?

Reply:
I think these case are already covered in the clause.

Reply:
Let me add tests and check it out.

Reply:
Added

Reply:
oops I missed the cases when structs are inside array & map, good point @dilipbiswal !